pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,277 @@
+	queryStats := stats.FromContext(ctx)
+	for _, req := range upstreamRequests {
+		r := req
+		g.Go(func() error {
+			partialStats, childCtx := stats.ContextWithEmptyStats(ctx)
+			partialStats.AddShardedQueries(1)
+
+			var span opentracing.Span
+			span, childCtx = opentracing.StartSpanFromContext(childCtx, "shardActiveSeries.doRequests")
+			defer span.Finish()
+
+			resp, err := next.RoundTrip(r.WithContext(childCtx))
+			if err != nil {
+				span.LogFields(otlog.Error(err))
+				return err
+			}
+
+			queryStats.Merge(partialStats)
+
+			mtx.Lock()
+			resps = append(resps, resp)
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	return resps, g.Wait()
+}
+
+func shardedSelector(shardCount, currentShard int, expr parser.Expr) (parser.Expr, error) {
+	originalSelector, ok := expr.(*parser.VectorSelector)
+	if !ok {
+		return nil, errors.New("invalid selector")
+	}
+
+	shardMatcher, err := labels.NewMatcher(
+		labels.MatchEqual, sharding.ShardLabel,
+		sharding.ShardSelector{ShardIndex: uint64(currentShard), ShardCount: uint64(shardCount)}.LabelValue(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &parser.VectorSelector{
+		Name:          originalSelector.Name,
+		LabelMatchers: append([]*labels.Matcher{shardMatcher}, originalSelector.LabelMatchers...),
+	}, nil
+}
+
+func (s *shardActiveSeriesMiddleware) mergeResponses(responses []*http.Response) *http.Response {
+	reader, writer := io.Pipe()
+
+	json = jsoniter.ConfigCompatibleWithStandardLibrary
+	items := make(chan any)
+
+	g := new(errgroup.Group)
+	for _, res := range responses {
+		r := res
+		g.Go(func() error {
+			defer func(Body io.ReadCloser) {
+				_ = Body.Close()
+			}(r.Body)
+
+			it := jsoniter.Parse(jsoniter.ConfigFastest, r.Body, 512)
+
+			field := it.ReadObject()
+			if field != "data" {
+				err := errors.New("expected data field at top level")
+				return err
+			}
+
+			if it.WhatIsNext() != jsoniter.ArrayValue {
+				err := errors.New("expected data field to contain an array")
+				return err
+			}
+
+			for it.ReadArray() {
+				items <- it.Read()
+			}
+
+			return it.Error
+		})
+	}
+
+	go s.writeMergedResponse(g, writer, items)
+
+	return &http.Response{Body: reader, StatusCode: http.StatusOK}
+}
+
+func (s *shardActiveSeriesMiddleware) writeMergedResponse(g *errgroup.Group, w io.WriteCloser, series chan any) {
+	defer func(w io.Closer) {
+		_ = w.Close()
+	}(w)
+
+	stream := jsoniter.NewStream(jsoniter.ConfigFastest, w, 512)
+	defer func(stream *jsoniter.Stream) {
+		_ = stream.Flush()
+	}(stream)
+
+	stream.WriteObjectStart()
+	stream.WriteObjectField("data")
+
+	stream.WriteArrayStart()
+
+	doneWriting := make(chan struct{})
+	go func() {
+		firstItem := true
+		for {
+			item, ok := <-series
+			if !ok {
+				doneWriting <- struct{}{}
+				return
+			}
+			if firstItem {
+				firstItem = false
+			} else {
+				stream.WriteMore()
+			}
+			stream.WriteVal(item)
+		}
+	}()
+
+	err := g.Wait()
+	close(series)
+	<-doneWriting
+
+	stream.WriteArrayEnd()
```

Comment:
This is cleaner and more concise, IMO: main flow is writing the array, the extra goroutine just controls the output.

```suggestion
	var err error
	go func() {
		err = g.Wait()
		close(series)
	}()
	
	stream.WriteObjectStart()
	stream.WriteObjectField("data")
	stream.WriteArrayStart()
	firstItem := true
	for item := range series {
		if firstItem {
			firstItem = false
		} else {
			stream.WriteMore()
		}
		stream.WriteVal(item)
	}
	stream.WriteArrayEnd()
```

I would also advocate for closing the channel in the same method where it's being written, like this:

```go
	go func() {
		// We ignore the error from the errgroup because it will be checked later again.
		_ = g.Wait()
		close(items)
	}()
	go s.writeMergedResponse(g.Wait, writer, items)

	return &http.Response{Body: reader, StatusCode: http.StatusOK}
}

func (s *shardActiveSeriesMiddleware) writeMergedResponse(check func() error, w io.WriteCloser, items chan any) {
	defer func(w io.Closer) {
		_ = w.Close()
	}(w)

	stream := jsoniter.NewStream(jsoniter.ConfigFastest, w, 512)
	defer func(stream *jsoniter.Stream) {
		_ = stream.Flush()
	}(stream)

	stream.WriteObjectStart()
	stream.WriteObjectField("data")
	stream.WriteArrayStart()
	firstItem := true
	for item := range items {
		if firstItem {
			firstItem = false
		} else {
			stream.WriteMore()
		}
		stream.WriteVal(item)
	}
	stream.WriteArrayEnd()

	if err := check(); err != nil {
		stream.WriteMore()
		stream.WriteObjectField("status")
		stream.WriteString("error")
		stream.WriteMore()
		stream.WriteObjectField("error")
		stream.WriteString(err.Error())
	}

	stream.WriteObjectEnd()
}
```
(Also note I renamed `series` to `items` to be consistent with naming everywhere.

Reply:
Thanks, that makes a lot of sense, changed.